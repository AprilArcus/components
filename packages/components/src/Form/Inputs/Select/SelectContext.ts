/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { createContext, Ref, RefObject, MutableRefObject } from 'react'
import { SelectData, SelectTransition } from './state'

export interface SelectContextProps {
  data: SelectData
  inputRef?: RefObject<HTMLInputElement>
  popoverRef?: Ref<HTMLElement>
  buttonRef?: Ref<HTMLButtonElement>
  onSelect?: (value: string) => void
  optionsRef?: MutableRefObject<string[]>
  state?: string
  transition?: SelectTransition
  listboxId?: string
  autocompletePropRef?: MutableRefObject<boolean>
  persistSelectionRef?: MutableRefObject<boolean>
  isVisible?: boolean
  openOnFocus?: boolean
}

export const defaultData = {
  // the value the user has navigated to with the keyboard
  navigationValue: undefined,
  // the value the user has typed, we derived this also when the developer is
  // controlling the value of SelectInput
  value: '',
}

export const SelectContext = createContext<SelectContextProps>({
  data: defaultData,
})

// Allows us to put the option's value on context so that SelectOptionText
// can work it's highlight text magic no matter what else is rendered around
// it.
export const OptionContext = createContext<string | undefined>(undefined)
