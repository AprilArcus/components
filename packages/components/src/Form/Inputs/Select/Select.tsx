/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, {
  forwardRef,
  Ref,
  RefObject,
  FormEvent,
  MouseEvent,
  useRef,
  useState,
} from 'react'
import { HotKeys } from 'react-hotkeys'
import styled from 'styled-components'
import {
  border,
  BorderProps,
  CompatibleHTMLProps,
  CustomizableAttributes,
  layout,
  LayoutProps,
  reset,
  space,
  SpaceProps,
  Theme,
  typography,
  TypographyProps,
} from '@looker/design-tokens'
import { ValidationType } from '../../ValidationMessage'
import { MenuList } from '../../../Menu'
import { usePopover } from '../../../Popover'
import { SelectContext } from './SelectContext'
import { useIndexNavigation } from '../../../utils/useIndexNavigation'

export const CustomizableSelectAttributes: CustomizableAttributes = {
  borderRadius: 'medium',
  fontSize: 'small',
  height: '28px',
  px: 'xsmall',
  py: 'none',
}

export interface SelectProps
  extends BorderProps,
    Omit<LayoutProps, 'size'>,
    SpaceProps,
    TypographyProps,
    Omit<CompatibleHTMLProps<HTMLDivElement>, 'onChange'> {
  /**
   * Use when Select is not controlled
   */
  defaultValue?: any
  /**
   * Displays an example value or short hint to the user. Should not replace a label.
   */
  placeholder?: string
  /**
   * Callback function fired when a SelectOption is selected
   */
  onChange?: (event: FormEvent<{ value: any }>) => void
  /**
   * Use value 'error' for error styling
   */
  validationType?: ValidationType
  /**
   * Use when Select is controlled
   */
  value?: any
}

const StyledInput = styled.input`
  ${reset}
  border: none;
  outline: none;
  background-color: transparent;
`

const SelectComponent = forwardRef(
  (
    {
      children,
      placeholder,
      defaultValue: propsDefault,
      value: propsValue,
      onChange,
      onClick,
      ...props
    }: SelectProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const [stateValue, setStateValue] = useState(propsDefault)
    const { index, gotoNext, gotoPrev } = useIndexNavigation()
    const innerRef = useRef<null | HTMLElement>(null)

    const menu = <MenuList>{children}</MenuList>

    const { popover, open, ref: triggerRef } = usePopover({
      arrow: false,
      content: menu,
    })

    function handleClick(event: MouseEvent<HTMLDivElement>) {
      open(event)
      onClick && onClick(event)
    }
    function handleChange(event: FormEvent<{ value: any }>) {
      onChange && onChange(event)
    }

    function handleTextChange(e: FormEvent<HTMLInputElement>) {
      setStateValue(e.currentTarget.value)
    }

    return (
      <SelectContext.Provider value={{ onChange: handleChange }}>
        <HotKeys
          innerRef={innerRef}
          keyMap={{ MOVE_DOWN: 'down', MOVE_UP: 'up' }}
          handlers={{
            MOVE_DOWN: () => moveFocus(1, 0, innerRef),
            MOVE_UP: () => moveFocus(-1, -1, innerRef),
          }}
        >
          <SelectBase
            borderColor="palette.charcoal300"
            display="inline-flex"
            onClick={handleClick}
            {...props}
            ref={triggerRef as RefObject<HTMLDivElement>}
          >
            <StyledInput
              type="text"
              readOnly
              value={propsValue || stateValue}
              placeholder={placeholder}
              onChange={handleTextChange}
              ref={ref}
            />
            {popover}
          </SelectBase>
        </HotKeys>
      </SelectContext.Provider>
    )
  }
)

SelectComponent.displayName = 'SelectComponent'

//
// @TODO - Should be properly imported from `Caret Down.svg`
// import caretDownIcon from '../../../../icons/svg/Caret Down.svg'
//
const indicatorRaw = `<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M7.41 8L12 12.58L16.59 8L18 9.41L12 15.41L6 9.41L7.41 8Z" fill="#1C2125"/>
</svg>`
const indicatorSize = '1rem'
const indicatorPadding = '.25rem'
const indicator = ({ theme }: { theme: Theme }) =>
  typeof window !== 'undefined' &&
  window.btoa(indicatorRaw.replace('#1C2125', theme.colors.palette.charcoal500))

const SelectBase = styled.div.attrs((props: SelectProps) => ({
  borderRadius: props.borderRadius || CustomizableSelectAttributes.borderRadius,
  fontSize: props.fontSize || CustomizableSelectAttributes.fontSize,
  height: props.py || props.p ? undefined : CustomizableSelectAttributes.height,
  px: props.p || CustomizableSelectAttributes.px,
  py: props.p || CustomizableSelectAttributes.py,
}))<SelectProps>`
  ${reset}
  display: inline-flex;
  align-items: stretch;

  background-color: ${props =>
    props.validationType === 'error'
      ? props.theme.colors.palette.red000
      : props.theme.colors.palette.white};
  border: solid 1px;

  background-image:
    url('data:image/svg+xml;base64,${indicator}'),
    linear-gradient(to bottom, white 0%, white 100%);

  background-repeat: no-repeat, repeat;
  background-position: right ${indicatorPadding} center, 0 0;
  background-size: ${indicatorSize}, 100%;

  ${border}
  ${layout}
  ${typography}
  ${space}
  padding-right: calc(2 * ${indicatorPadding} + ${indicatorSize});

`

export const Select = styled(SelectComponent)``
