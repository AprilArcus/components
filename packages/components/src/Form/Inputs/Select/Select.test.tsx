/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import { mountWithTheme, assertSnapshot } from '@looker/components-test-utils'

import { Select } from './Select'

test('Select default', () => {
  assertSnapshot(<Select id="default" />)
})

test('Select with name and id', () => {
  assertSnapshot(<Select name="Bob" id="Bobby" />)
})

test('Select should accept disabled', () => {
  assertSnapshot(<Select disabled id="disabled" />)
})

test('Select should accept empty options array', () => {
  assertSnapshot(<Select options={[]} id="empty-options" />)
})

test('Select with a placeholder', () => {
  assertSnapshot(<Select placeholder="I am a placeholder" id="placeholder" />)
})

test('Select placeholder option does not have a value', () => {
  const select = mountWithTheme(<Select placeholder="Boo!" id="no-value" />)
  expect(select.find('input').prop('value')).toEqual('')
})

test('Select should accept readOnly', () => {
  assertSnapshot(<Select readOnly id="read-only" />)
})

test('Select should accept required', () => {
  assertSnapshot(<Select required id="required" />)
})

test('Select with a value', () => {
  const options = [
    { data: '1', value: 'thing' },
    { data: '1.5', value: 'Some Value' },
    { data: '2', value: 'other' },
  ]
  assertSnapshot(
    <Select value="Some Value" options={options} id="value-and-options" />
  )
})

test('Select with aria-describedby', () => {
  assertSnapshot(<Select aria-describedby="some-id" />)
})

test('Select with an error validation', () => {
  assertSnapshot(<Select validationType="error" id="error" />)
})

test('Should trigger onChange handler', () => {
  const handleChange = jest.fn()

  const wrapper = mountWithTheme(
    <Select
      openOnFocus
      onSelect={handleChange}
      options={[{ value: 'foo' }, { value: 'bar' }]}
    />
  )

  wrapper.find('input').simulate('focus')
  wrapper.update()
  wrapper
    .find('li')
    .at(0)
    .simulate('click')
  expect(handleChange).toHaveBeenCalledTimes(1)
  expect(handleChange).toHaveBeenCalledWith({ value: 'foo' })
})
