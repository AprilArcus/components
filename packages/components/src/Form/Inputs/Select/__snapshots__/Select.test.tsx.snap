// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select default 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-default"
  className="c0"
  id="default"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select should accept disabled 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-disabled"
  className="c0"
  disabled={true}
  id="disabled"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-disabled"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select should accept empty options array 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-empty-options"
  className="c0"
  id="empty-options"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-empty-options"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select should accept readOnly 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-read-only"
  className="c0"
  id="read-only"
  readOnly={true}
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-read-only"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select should accept required 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-required"
  className="c0"
  id="required"
  required={true}
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-required"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select with a placeholder 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-placeholder"
  className="c0"
  id="placeholder"
  placeholder="I am a placeholder"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-placeholder"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select with a value 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-value-and-options"
  className="c0"
  id="value-and-options"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-value-and-options"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value="Some Value"
  />
</div>
`;

exports[`Select with an error validation 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-error"
  className="c0"
  id="error"
  role="select"
  validationType="error"
>
  <input
    className="c1 c2"
    id="listbox-error"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select with aria-describedby 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-describedby="some-id"
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-3d263117-ba3a-40a8-863c-06a910a6d1be"
  className="c0"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-3d263117-ba3a-40a8-863c-06a910a6d1be"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`Select with name and id 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: solid 1px;
  border-color: #C1C6CC;
  border-radius: 0.25rem;
  width: 100%;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c2 {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTcuNDEgOEwxMiAxMi41OEwxNi41OSA4TDE4IDkuNDFMMTIgMTUuNDFMNiA5LjQxTDcuNDEgOFoiIGZpbGw9IiM3MDc3ODEiLz4KPC9zdmc+');
  background-repeat: no-repeat,repeat;
  background-position: right .25rem center,0 0;
  background-size: 1rem,100%;
  padding-right: calc(2 * .25rem + 1rem);
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="listbox-Bobby"
  className="c0"
  id="Bobby"
  name="Bob"
  role="select"
>
  <input
    className="c1 c2"
    id="listbox-Bobby"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;
